lib/pbench/agent/tool_data_sink.py

```diff
@@ -1295,13 +1300,29 @@ def execute(self):
                 )
             self.logger.debug("published %s", self._to_client_channel)
 
-            for data in self._from_client_chan.fetch_json(self.logger):
-                ret_val = self._valid_data(data)
-                if ret_val is None:
+            # Begin listening for client state signals, and react/respond
+            # accordingly upon signal receipt
+            for signal in self.sig_resp.listen():
+                client = signal.publisher_id
+                action = signal.event
+                if action == "shutdown":
+                    # This simply means that a SignalExporter/Client we
+                    # were listening to has shut down, no action needed
+                    continue
+                if action == "initialization":
+                    self.sig_resp.srespond(signal)
                     continue
-                action, directory, args = ret_val[0], ret_val[1], ret_val[2]
-                self.execute_action(action, directory, args, data)
-            self.logger.info("Tool Data Sink terminating")
+                if not signal.metadata:
```

Comment:
What does it mean to get a signal without metadata?  Perhaps we should add a `self.logger.debug()` with the contents?